alf/algorithms/ppo_loss_test.py

```diff
@@ -0,0 +1,121 @@
+from tf_agents.networks.value_network import ValueNetwork
+from tf_agents.networks.value_rnn_network import ValueRnnNetwork
+
+from alf.algorithms.actor_critic_algorithm import ActorCriticAlgorithm
+from alf.algorithms.on_policy_algorithm import OffPolicyAdapter
+from alf.algorithms.ppo_loss import PPOLoss
+from alf.drivers.off_policy_driver import OffPolicyDriver
+from alf.environments.suite_unittest import PolicyUnittestEnv
+from alf.environments.suite_unittest import ActionType
+
+
+def create_algorithm(env, use_rnn=False, learning_rate=1e-1):
+    observation_spec = env.observation_spec()
+    action_spec = env.action_spec()
+
+    if use_rnn:
+        actor_net = ActorDistributionRnnNetwork(
+            observation_spec,
+            action_spec,
+            input_fc_layer_params=(),
+            output_fc_layer_params=(),
+            lstm_size=(4, ))
+        value_net = ValueRnnNetwork(
+            observation_spec,
+            input_fc_layer_params=(),
+            output_fc_layer_params=(),
+            lstm_size=(4, ))
+    else:
+        actor_net = ActorDistributionNetwork(
+            observation_spec, action_spec, fc_layer_params=())
+        value_net = ValueNetwork(observation_spec, fc_layer_params=())
+
+    optimizer = tf.optimizers.Adam(learning_rate=learning_rate)
+
+    ac_algorithm = ActorCriticAlgorithm(
+        action_spec=action_spec,
+        actor_network=actor_net,
+        value_network=value_net,
+        loss=PPOLoss(action_spec=action_spec, gamma=1.0),
+        optimizer=optimizer)
+    return OffPolicyAdapter(ac_algorithm)
+
+
+class PpoTest(unittest.TestCase):
+    def test_ppo(self):
+        batch_size = 100
+        steps_per_episode = 13
+        env = PolicyUnittestEnv(batch_size, steps_per_episode)
+        env = TFPyEnvironment(env)
+
+        eval_env = PolicyUnittestEnv(batch_size, steps_per_episode)
+        eval_env = TFPyEnvironment(eval_env)
+
+        algorithm = create_algorithm(env)
+        driver = OffPolicyDriver(
+            env,
+            algorithm,
+            debug_summaries=True,
+            summarize_grads_and_vars=True)
+        replay_buffer = driver.add_replay_buffer()
+        eval_driver = OffPolicyDriver(
+            eval_env, algorithm, training=False, greedy_predict=True)
+        driver.run = tf.function(driver.run)
+        eval_driver.run = tf.function(eval_driver.run)
+
+        env.reset()
+        eval_env.reset()
+        time_step = driver.get_initial_time_step()
+        policy_state = driver.get_initial_state()
+        for i in range(20):
+            time_step, policy_state = driver.run(
+                max_num_steps=batch_size * steps_per_episode,
+                time_step=time_step,
+                policy_state=policy_state)
+
+            experience = replay_buffer.gather_all()
+            driver.train(experience, mini_batch_size=25)
+            replay_buffer.clear()
```

Comment:
Why using an off-policy scheme and clearing replay buffer each time, rather than implementing PPO with on-policy driver?

Reply:
Because algorithms such as PPO needs to go through the data multiple times.